pkg/kstatus/watcher/watcher_dynamic.go

```diff
@@ -0,0 +1,825 @@
+		ObjectFilter: &AllowListFilter{
+			AllowList: ids,
+		},
+	}
+	return informer.Start(ctx)
+}
+
+// ObjectStatusInformer watches a cluster for object status updates.
+//
+// ObjectStatusInformer supports two dimensions of filtering:
+// 1. Only the specified GroupKinds are watched.
+// 2. Objects can be ignored using the supplied filter.
+//
+// ObjectStatusInformer is NOT repeatable. It will panic if started more than
+// once. If you need a repeatable solution, try DynamicStatusWatcher.
+type ObjectStatusInformer struct {
+	// DynamicClient is used to watch of resource objects.
+	DynamicClient dynamic.Interface
+
+	// Mapper is used to map from GroupKind to GroupVersionKind.
+	Mapper meta.RESTMapper
+
+	// ResyncPeriod is how often the objects are retrieved to re-synchronize,
+	// in case any events were missed.
+	ResyncPeriod time.Duration
+
+	// StatusReader specifies a custom implementation of the
+	// engine.StatusReader interface that will be used to compute reconcile
+	// status for resource objects.
+	StatusReader engine.StatusReader
+
+	// ClusterReader is used to look up generated objects on-demand.
+	// Generated objects (ex: Deployment > ReplicaSet > Pod) are sometimes
+	// required for computing parent object status, to compensate for
+	// controllers that aren't following status conventions.
+	ClusterReader engine.ClusterReader
+
+	// GroupKinds is the list of GroupKinds to watch.
+	GroupKinds []schema.GroupKind
+
+	// ObjectFilter is used to decide which objects to ingore.
+	ObjectFilter ObjMetadataFilter
+
+	// lock guards the subsequent stateful fields
+	lock sync.Mutex
+
+	// pendingWatchers tracks GroupKinds of custom resources which are not yet
+	// registered with the cluster.
+	// Once a CRD is created, an informer will be created to watch it.
+	pendingWatchers *groupKindSet
+
+	// delayedEventCanceller makes it possible to cancel delayed events.
+	// Unschedulable pods cause a delayed event after 15s, if not cancelled.
+	delayedEventCanceller *objectCanceller
+
+	// watchCanceller is used to stop informers for specific GroupKinds.
+	watchCanceller *groupKindCanceller
+
+	// muxer is used to get a single output event channel for all informers.
+	// New input channels may be added at runtime as informers are created.
+	muxer *eventFunnel
+
+	started bool
+	stopped bool
+}
+
+func (w *ObjectStatusInformer) Start(ctx context.Context) <-chan event.Event {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	if w.started {
+		panic("ObjectStatusInformer cannot be restarted")
+	}
+
+	w.pendingWatchers = &groupKindSet{}
+	w.delayedEventCanceller = &objectCanceller{}
+	w.watchCanceller = &groupKindCanceller{}
+	w.muxer = &eventFunnel{}
+
+	eventCh := make(chan event.Event)
+	w.muxer.Input(eventCh)
+
+	// Build & Start an informer for each unique GroupKind.
+	for _, groupKind := range w.GroupKinds {
+		err := w.startForGroupKind(
+			ctx,
+			groupKind,
+		)
+		if err != nil {
+			handleError(eventCh, err)
```

Comment:
Do we keep running here even if we fail to start the informer for a GroupKind? Unless it has changed recently, the expectation from clients should be that an ErrorEvent is used for fatal errors and will be followed by the event channel being closed.

Reply:
Good catch. I'll make it return. It looks like that's what the StatusPoller does for every error.

Reply:
Done. All error events are now terminal.